### DISCUSSION TITLE
Support for 'jalt' feature and target length

### DISCUSSION THREAD

**Discussion by myoresh** [harfbuzz/harfbuzz#3252] [Ideas] [answered]
Would it be possible to support 'jalt' OT feature in HarfBuzz? This feature is very useful in traditional Hebrew typesetting, but could be applicable to other scripts as well. For example, Microsoft argues for its use in Arabic (see https://docs.microsoft.com/en-us/typography/opentype/spec/features_fj#tag-jalt).

The problem with justification, as I see it, is that not all characters with alternative-width substitutions should be automatically replaced. In fact, the replacements should be made one after another, until the desired width is reached.

Is something like that already achievable in HarfBuzz or can be implemented with reasonable effort?

**Comment by behdad** [comment, marked as answer]
HarfBuzz currently does not implement justification, because it's a hard problem, and is under-specified in OpenType.

**Reply by myoresh** [reply]
Would it be feasible to do some steps towards the solution of this problem? Justification, especially in complex text layout, is an issue faced by many users of HarfBuzz API, and they all solve it with more or less the same approaches. Maybe we can step into this void and provide the de-facto solution. I can contribute to it, as necessary. It looks like most justification issues can be covered by application of things HarfBuzz already knows how to do.

**Reply by behdad** [reply]
Trust me when I say I share your passion about this problem :).
http://behdad.org/urdu/

Sure, let's figure out what we can do.

**Reply by behdad** [reply]
Note that we have `hb_ot_layout_lookup_get_glyph_alternates()`. But that's really not that much help.

**Reply by myoresh** [reply]
What do you think, what would be the minimal functionality worth releasing? I can think of sort of fallback support for the usual Latin-style justification with OpenType-based stuff on top of that. The OpenType justification can be generic or script-specific or a blend of both. I'm afraid that without the Latin part, the users won't bother to use it at all.

Could you please also draft the desired API extension? I'm not quite accustomed with it.

**Reply by myoresh** [reply]
Please take a look at the draft API: https://github.com/harfbuzz/harfbuzz/pull/3259
